src/fields/unit_converter_template.hh

```diff
@@ -0,0 +1,262 @@
+    }
+
+
+    UnitData unit_data_;         //!< Full parsed data
+    std::string unit_data_key_;  //!< keyo actual item of unit_data_
+    int factor_idx_;             //!< index to actual item of subvector of factors of unit_data_
+};
+
+
+// the spirit grammer
+template< class Iter_type >
+class UnitSIGrammer : public spirit_namespace::grammar< UnitSIGrammer< Iter_type > >
+{
+public:
+
+    typedef Semantic_actions< Iter_type > Semantic_actions_t;
+
+    UnitSIGrammer( Semantic_actions_t& semantic_actions )
+    :   actions_( semantic_actions )
+    {
+    }
+
+	static void throw_not_equating( Iter_type begin, Iter_type end )
+    {
+		THROW( ExcInvalidUnit() << EI_UnitError("Invalid expression, missing '='") );
+    }
+
+	static void throw_exp_not_int( Iter_type begin, Iter_type end )
+    {
+		THROW( ExcInvalidUnit() << EI_UnitError("Value of exponent is not integer") );
+    }
+
+
+	static void throw_not_shortcut( Iter_type begin, Iter_type end )
+    {
+		THROW( ExcInvalidUnit() << EI_UnitError("Invalid shortcut of unit") );
+    }
+
+
+    template< typename ScannerT >
+    class definition
+    {
+    public:
+
+        definition( const UnitSIGrammer& self )
+        {
+            using namespace spirit_namespace;
+
+            // first we convert the semantic action class methods to functors with the
+            // parameter signature expected by spirit
+
+            typedef boost::function< void( Iter_type, Iter_type ) > Str_action;
+            typedef boost::function< void( double )               > Real_action;
+            typedef boost::function< void( boost::int64_t )       > Int_action;
+
+            Str_action    new_shortcut     ( boost::bind( &Semantic_actions_t::new_shortcut,     &self.actions_, _1, _2 ) );
+            Str_action    new_mult_factor  ( boost::bind( &Semantic_actions_t::new_mult_factor,  &self.actions_, _1, _2 ) );
+            Str_action    new_div_factor   ( boost::bind( &Semantic_actions_t::new_div_factor,   &self.actions_, _1, _2 ) );
+            Real_action   new_multipl      ( boost::bind( &Semantic_actions_t::new_multipl,      &self.actions_, _1 ) );
+            Int_action    new_exp          ( boost::bind( &Semantic_actions_t::new_exp,          &self.actions_, _1 ) );
+
+            // actual grammer
+
+            unit_
+			    = formula_ >> *( ch_p(';') >> *space_p >> constant_ )
+                ;
+
+            formula_
+                = formula_factor_[ new_mult_factor ] >> !( ch_p('^') >> exp_ )
+                  >> *( (ch_p('*') >> formula_factor_[ new_mult_factor ] >> !( ch_p('^') >> exp_ ) )
+                      | (ch_p('/') >> formula_factor_[ new_div_factor ] >> !( ch_p('^') >> exp_ ) )
+                      )
+                ;
+
+            formula_factor_
+		        = +alpha_p | eps_p[ &throw_not_shortcut ]
+			    ;
+
+            constant_
+			    = constant_shortcut_[ new_shortcut ]
+				  >> *space_p >> ( ch_p('=') | eps_p[ &throw_not_equating ] )
```

Comment:
Try to implement  information about position in error messages, e.g
http://www.boost.org/doc/libs/1_61_0/libs/spirit/classic/example/fundamental/position_iterator/position_iterator.cpp